Implemented subscriber manager

oopsmonk opened the pull request:
This PR fixes #24 
The subscriber manager manages topics of subscribers based on connect/disconnect/subscribe/unsubscribe events

Commit:
added subscriber manager

TODO: unsubcribe on disconnection

Commit:
Unsubscribe topics on disconnection

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+	"sync"
+)
+
+type OnSubscribeH func(topic string, id string)
```
Call them `Func` at the end.
That fits better to the rest of our codebase.

muXxer commented on core/mqtt/server.go:
```diff
@@ -190,6 +196,18 @@ func (s *Server) onUnsubscribeTopic(topic string) {
+
+func (s *Server) onClientDisconnect(id string) {
+	// unsubscribe topics of the client which it might lost connection accidentally
+	s.LogDebug("onClientDisconnect")
```
Maybe also add the client ID in the Debug log like in the other logs.

muXxer commented on core/mqtt/server.go:
```diff
@@ -190,6 +196,18 @@ func (s *Server) onUnsubscribeTopic(topic string) {
 }
 
+func (s *Server) onClientConnect(id string) {
+	// nothing to do at the moment
```
Also add a debug log here

muXxer commented on core/mqtt/server.go:
```diff
@@ -123,7 +127,8 @@ func (s *Server) Run(ctx context.Context) {
 
-func (s *Server) onSubscribeTopic(ctx context.Context, topic string) {
+func (s *Server) onSubscribeTopic(ctx context.Context, id string, topic string) {
+	s.LogDebugf("%s subscriber: %s", id, topic)
```
```suggestion
	s.LogDebugf("%s subscribed to topic %s", id, topic)
```

muXxer commented on core/mqtt/server.go:
```diff
@@ -165,7 +170,8 @@ func (s *Server) onSubscribeTopic(ctx context.Context, topic string) {
 
-func (s *Server) onUnsubscribeTopic(topic string) {
+func (s *Server) onUnsubscribeTopic(id string, topic string) {
+	s.LogDebugf("%s unsubscribe: %s", id, topic)
```
```suggestion
	s.LogDebugf("%s unsubscribed from %s", id, topic)
```

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+type OnConnectH func(id string)
+type OnDisconnectH func(id string)
+
+type subscriberManager struct {
```
I would call it `subscriptionManager`.

oopsmonk replied:
it stores subscriber information for server handling subscriptions. the subscription management logic is implemented in the server.go, I'd like to keep it as `subscriberManager`.

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+
+type subscriberManager struct {
+	// a map keeps client ID and topics
+	subscribers    map[string]map[string]string
```
Maybe call it `subscriptions` because it is not only subscribers, but also their topics.

oopsmonk replied:
call it subscribers is more specific, a subscriber has its own topics and a subscription is a collection of topics and subscribers, imo.

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// add the client ID to map
+	s.subscribers[id] = make(map[string]string)
```
Maybe we should check if the client already exists?

oopsmonk replied:
good catch!
it's more complex than just checking the existence, it depends on the mqtt client implementation, will figure out what's a proper way to deal with it later.

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+func (s *subscriberManager) Subscribe(id string, topicName string) {
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// add the topic to a crrosponding ID
```
typo

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// add the topic to a crrosponding ID
+	s.subscribers[id][topicName] = topicName
```
better check if the client ID already exists, and create it if not.

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+func (s *subscriberManager) Unsubscribe(id string, topicName string) {
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// remove the topic from a crrosponding ID
```
typo

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+package mqtt
```
Maybe it would make sense to create test cases for the subscriptionManager.

muXxer commented on pkg/mqtt/broker.go:
```diff
@@ -14,13 +14,13 @@ import (
-	topicManager *topicManager
+	broker      *mqtt.Server
+	opts        *BrokerOptions
+	subMananger *subscriberManager
```
`Mananger` is a typo.
I also would write a longer name, because "sub" doesn't ring a bell regarding subscriptions in my head :)

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+
+// Returns topis of a subscriber
+func (s *subscriberManager) Topics(id string) map[string]string {
+	return s.subscribers[id]
```
I don't think it is a good idea to return the map, without holding the lock.
Maybe create an `ForEachTopic` instead, which gets a consumer function passed.

oopsmonk replied:
it's a get method which likely does not have race condition issues. the returned map is a copy.

alexsporn replied:
Map types are reference types, so there is no implicit copy here

muXxer commented on pkg/mqtt/topic_manager.go:
```diff
@@ -1,101 +0,0 @@
-	t.subscribedTopicsLock.Lock()
-	defer t.subscribedTopicsLock.Unlock()
-
-	count, has := t.subscribedTopics[topicName]
```
Why did you remove that whole logic with the counters?
This is essential in my opinion.

oopsmonk replied:
the subscription counters are implemented in server.go. the topic manager is replaced by the subscriber manager, this counter is not necessary for subscriber manager.

muXxer commented on pkg/mqtt/topic_manager.go:
```diff
@@ -1,101 +0,0 @@
-
-// cleanupWithoutLocking recreates the subscribedTopics map to release memory for the garbage collector.
-func (t *topicManager) cleanupWithoutLocking() {
-	subscribedTopics := make(map[string]int)
```
Why did you remove the cleanup code?
Also essential.

oopsmonk replied:
ditto

muXxer replied:
This was added on purpose, because the map memory leaks, because the garbage collector never cleans it up.
This is a go specific problem.

oopsmonk replied:
I'd like to learn more about the GC issue with the map object, could you provide some materials? thanks in advance.

muXxer commented on pkg/mqtt/subscriber_manager.go:
```diff
@@ -0,0 +1,106 @@
+type OnConnectH func(id string)
+type OnDisconnectH func(id string)
+
+type subscriberManager struct {
```
Add the description what the subscriptionManager does (take the one from the topic manager and extend it)

Commit:
addressed comments

- renamed callback functions
- fixes typos

TODO: handle dupolicate client, add test cases

Commit:
typo

Commit:
fixed duplicate client issue

added test cases for subscriber manager

Commit:
fulfil static check requests

(S1002) should omit comparison to bool constant
(ST1005) error strings should not be capitalized

Commit:
Update test cases

Commit:
maps cleanup for the garbage collector

Commit:
assign capacity in map cleanup

Commit:
Use ShrinkingMap instead of map

Commit:
Update shrinkingmap

muXxer committed:
Fix subscription logic and adapt to new MQTT Broker version (#30)

* Fix test cases

* Move subscription logic to SubscriptionManager

* Adapt TLS Config to new mqtt broker version

* Release 0.10.1
